// @visa/charts/dist/collection/components/dumbbell-plot/dumbbell-plot-default-values.js
/**
 * Copyright (c) 2020, 2021 Visa, Inc.
 *
 * This source code is licensed under the MIT license
 * https://github.com/visa/visa-chart-components/blob/master/LICENSE
 *
 **/
import Utils from '@visa/visa-charts-utils';
const { propDefaultValues } = Utils;
export class DumbbellPlotDefaultValues {
}
DumbbellPlotDefaultValues.mainTitle = 'Dumbbell Plot Title';
DumbbellPlotDefaultValues.subTitle = "This is the dumbbell plot's subtitle";
DumbbellPlotDefaultValues.height = 300;
DumbbellPlotDefaultValues.width = 650;
DumbbellPlotDefaultValues.margin = {
  top: 300 * 0.01,
  bottom: 300 * 0.01,
  right: 650 * 0.01,
  left: 650 * 0.01
};
DumbbellPlotDefaultValues.padding = {
  top: 20,
  bottom: 50,
  right: 50,
  left: 50
};
DumbbellPlotDefaultValues.ordinalAccessor = 'date';
DumbbellPlotDefaultValues.valueAccessor = 'value';
DumbbellPlotDefaultValues.seriesAccessor = 'category';
// dumbbell specific props
DumbbellPlotDefaultValues.focusMarker = {
  key: '',
  sizeFromBar: 12
};
DumbbellPlotDefaultValues.marker = {
  visible: true,
  type: 'dot',
  sizeFromBar: 8
};
DumbbellPlotDefaultValues.barStyle = {
  opacity: 1,
  width: 1,
  colorRule: 'default'
};
DumbbellPlotDefaultValues.differenceLabel = {
  visible: false,
  placement: 'left',
  calculation: 'difference',
  format: '0[.][0][0]a',
  collisionHideOnly: false
};
DumbbellPlotDefaultValues.layout = propDefaultValues.layout;
DumbbellPlotDefaultValues.accessibility = propDefaultValues.accessibility;
DumbbellPlotDefaultValues.animationConfig = propDefaultValues.animationConfig;
DumbbellPlotDefaultValues.highestHeadingLevel = propDefaultValues.highestHeadingLevel;
DumbbellPlotDefaultValues.sortOrder = propDefaultValues.sortOrder;
DumbbellPlotDefaultValues.xAxis = propDefaultValues.xAxisDate;
DumbbellPlotDefaultValues.yAxis = propDefaultValues.yAxis;
DumbbellPlotDefaultValues.showBaselineX = propDefaultValues.showBaselineFalse;
DumbbellPlotDefaultValues.showBaselineY = propDefaultValues.showBaselineFalse;
DumbbellPlotDefaultValues.wrapLabel = propDefaultValues.wrapLabel;
DumbbellPlotDefaultValues.colorPalette = propDefaultValues.colorPaletteCategorical;
DumbbellPlotDefaultValues.hoverStyle = propDefaultValues.hoverStyle;
DumbbellPlotDefaultValues.clickStyle = propDefaultValues.clickStyle;
DumbbellPlotDefaultValues.referenceStyle = propDefaultValues.referenceStyle;
DumbbellPlotDefaultValues.cursor = propDefaultValues.cursor;
DumbbellPlotDefaultValues.roundedCorner = propDefaultValues.roundedCorner;
DumbbellPlotDefaultValues.barIntervalRatio = propDefaultValues.barIntervalRatio;
DumbbellPlotDefaultValues.hoverOpacity = propDefaultValues.hoverOpacity;
DumbbellPlotDefaultValues.dataLabel = propDefaultValues.dataLabelEnds;
DumbbellPlotDefaultValues.legend = propDefaultValues.hiddenLegend;
DumbbellPlotDefaultValues.tooltipLabel = propDefaultValues.tooltipLabel;
DumbbellPlotDefaultValues.showTooltip = propDefaultValues.showTooltip;
DumbbellPlotDefaultValues.suppressEvents = propDefaultValues.suppressEvents;
DumbbellPlotDefaultValues.referenceLines = propDefaultValues.referenceLines;
DumbbellPlotDefaultValues.annotations = propDefaultValues.annotations;
DumbbellPlotDefaultValues.clickHighlight = propDefaultValues.clickHighlight;
DumbbellPlotDefaultValues.seriesLabel = propDefaultValues.seriesLabel;
