// @visa/charts/dist/collection/components/alluvial-diagram/alluvial-diagram-default-values.js
/**
 * Copyright (c) 2021 Visa, Inc.
 *
 * This source code is licensed under the MIT license
 * https://github.com/visa/visa-chart-components/blob/master/LICENSE
 *
 **/
import Utils from '@visa/visa-charts-utils';
const { propDefaultValues } = Utils;
export class AlluvialDiagramDefaultValues {
}
AlluvialDiagramDefaultValues.highestHeadingLevel = propDefaultValues.highestHeadingLevel;
AlluvialDiagramDefaultValues.mainTitle = 'Alluvial Diagram Title';
AlluvialDiagramDefaultValues.subTitle = "This is the alluvial diagram's subtitle";
AlluvialDiagramDefaultValues.height = 300;
AlluvialDiagramDefaultValues.width = 650;
AlluvialDiagramDefaultValues.margin = {
  top: 300 * 0.01,
  bottom: 300 * 0.01,
  right: 650 * 0.01,
  left: 650 * 0.01
};
AlluvialDiagramDefaultValues.padding = {
  top: 20,
  bottom: 50,
  right: 50,
  left: 50
};
AlluvialDiagramDefaultValues.sourceAccessor = 'source';
AlluvialDiagramDefaultValues.targetAccessor = 'target';
AlluvialDiagramDefaultValues.valueAccessor = 'value';
AlluvialDiagramDefaultValues.groupAccessor = '';
// static readonly labelAccessor = '';
AlluvialDiagramDefaultValues.nodeIDAccessor = 'id';
AlluvialDiagramDefaultValues.nodeConfig = {
  fill: false,
  width: 12,
  padding: 20,
  alignment: 'left',
  compare: false
};
AlluvialDiagramDefaultValues.linkConfig = {
  visible: true,
  fillMode: 'none',
  opacity: 0.3
};
AlluvialDiagramDefaultValues.accessibility = propDefaultValues.accessibility;
AlluvialDiagramDefaultValues.animationConfig = propDefaultValues.animationConfig;
AlluvialDiagramDefaultValues.colorPalette = propDefaultValues.colorPaletteCategorical;
AlluvialDiagramDefaultValues.dataLabel = propDefaultValues.dataLabel;
AlluvialDiagramDefaultValues.annotations = propDefaultValues.annotations;
AlluvialDiagramDefaultValues.hoverStyle = propDefaultValues.hoverStyle;
AlluvialDiagramDefaultValues.clickStyle = propDefaultValues.clickStyle;
AlluvialDiagramDefaultValues.hoverOpacity = propDefaultValues.hoverOpacity;
AlluvialDiagramDefaultValues.cursor = propDefaultValues.cursor;
AlluvialDiagramDefaultValues.tooltipLabel = propDefaultValues.tooltipLabel;
AlluvialDiagramDefaultValues.showTooltip = propDefaultValues.showTooltip;
AlluvialDiagramDefaultValues.suppressEvents = propDefaultValues.suppressEvents;
AlluvialDiagramDefaultValues.clickHighlight = propDefaultValues.clickHighlight;
