// @visa/charts/dist/collection/components/world-map/world-map-default-values.js
/**
 * Copyright (c) 2020, 2021 Visa, Inc.
 *
 * This source code is licensed under the MIT license
 * https://github.com/visa/visa-chart-components/blob/master/LICENSE
 *
 **/
import Utils from '@visa/visa-charts-utils';
const { propDefaultValues } = Utils;
export class WorldMapDefaultValues {
}
// base, margin and padding
WorldMapDefaultValues.highestHeadingLevel = propDefaultValues.highestHeadingLevel;
WorldMapDefaultValues.mainTitle = 'This is the Map title';
WorldMapDefaultValues.subTitle = 'This is the Map subtitle';
WorldMapDefaultValues.height = 400;
WorldMapDefaultValues.width = 500;
WorldMapDefaultValues.margin = {
  top: 400 * 0.02,
  bottom: 400 * 0.02,
  right: 500 * 0.02,
  left: 500 * 0.02
};
WorldMapDefaultValues.padding = {
  top: 400 * 0.02,
  bottom: 400 * 0.02,
  right: 500 * 0.01,
  left: 500 * 0.01
};
// map
WorldMapDefaultValues.mapProjection = 'Equal Earth';
WorldMapDefaultValues.mapScaleZoom = 1.25;
WorldMapDefaultValues.quality = 'High';
// accessors
WorldMapDefaultValues.markerAccessor = '';
WorldMapDefaultValues.markerNameAccessor = '';
WorldMapDefaultValues.joinAccessor = 'countryCode';
WorldMapDefaultValues.joinNameAccessor = 'countryName';
WorldMapDefaultValues.valueAccessor = 'value';
WorldMapDefaultValues.latitudeAccessor = '';
WorldMapDefaultValues.longitudeAccessor = '';
WorldMapDefaultValues.groupAccessor = '';
// accessibility
WorldMapDefaultValues.accessibility = propDefaultValues.accessibility;
WorldMapDefaultValues.animationConfig = propDefaultValues.animationConfig;
// style
WorldMapDefaultValues.showGridLines = false;
WorldMapDefaultValues.sortOrder = propDefaultValues.sortOrder;
WorldMapDefaultValues.colorPalette = propDefaultValues.colorPaletteSequential;
WorldMapDefaultValues.colorSteps = propDefaultValues.colorSteps4;
WorldMapDefaultValues.showTooltip = propDefaultValues.showTooltip;
WorldMapDefaultValues.hoverOpacity = propDefaultValues.hoverOpacity;
WorldMapDefaultValues.markerStyle = {
  visible: false,
  blend: false,
  fill: true,
  radius: 5,
  radiusRange: '',
  opacity: 1,
  color: 'base_grey',
  strokeWidth: 1
};
WorldMapDefaultValues.countryStyle = {
  fill: true,
  opacity: 0.8,
  color: 'base_grey',
  strokeWidth: 0.5
};
WorldMapDefaultValues.hoverStyle = propDefaultValues.hoverStyle;
WorldMapDefaultValues.clickStyle = propDefaultValues.clickStyle;
WorldMapDefaultValues.cursor = propDefaultValues.cursor;
WorldMapDefaultValues.suppressEvents = propDefaultValues.suppressEvents;
WorldMapDefaultValues.dataLabel = propDefaultValues.hiddenDataLabel;
WorldMapDefaultValues.legend = propDefaultValues.keyLegend;
WorldMapDefaultValues.tooltipLabel = propDefaultValues.tooltipLabel;
WorldMapDefaultValues.annotations = propDefaultValues.annotations;
WorldMapDefaultValues.clickHighlight = propDefaultValues.clickHighlight;
