// @visa/charts/dist/cjs/pow-2996236d.js
'use strict';

const linear = require('./linear-aa74902e.js');

function transformPow(exponent) {
  return function(x) {
    return x < 0 ? -Math.pow(-x, exponent) : Math.pow(x, exponent);
  };
}

function transformSqrt(x) {
  return x < 0 ? -Math.sqrt(-x) : Math.sqrt(x);
}

function transformSquare(x) {
  return x < 0 ? -x * x : x * x;
}

function powish(transform) {
  var scale = transform(linear.identity, linear.identity),
      exponent = 1;

  function rescale() {
    return exponent === 1 ? transform(linear.identity, linear.identity)
        : exponent === 0.5 ? transform(transformSqrt, transformSquare)
        : transform(transformPow(exponent), transformPow(1 / exponent));
  }

  scale.exponent = function(_) {
    return arguments.length ? (exponent = +_, rescale()) : exponent;
  };

  return linear.linearish(scale);
}

function pow() {
  var scale = powish(linear.transformer());

  scale.copy = function() {
    return linear.copy(scale, pow()).exponent(scale.exponent());
  };

  linear.initRange.apply(scale, arguments);

  return scale;
}

function sqrt() {
  return pow.apply(null, arguments).exponent(0.5);
}

exports.pow = pow;
exports.powish = powish;
exports.sqrt = sqrt;
